ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java

```diff
@@ -126,1232 +130,1243 @@
 import org.antlr.v4.runtime.tree.TerminalNode;
 import org.apache.kafka.connect.data.Field;
 
-// CHECKSTYLE_RULES.OFF: ClassDataAbstractionCoupling
-public class AstBuilder extends SqlBaseBaseVisitor<Node> {
-  // CHECKSTYLE_RULES.ON: ClassDataAbstractionCoupling
+public class AstBuilder {
 
-  private int selectItemIndex = 0;
+  private final MetaStore metaStore;
 
-  private static final String DEFAULT_WINDOW_NAME = "StreamWindow";
+  AstBuilder(final MetaStore metaStore) {
+    this.metaStore = requireNonNull(metaStore, "metaStore");
+  }
 
-  private final DataSourceExtractor dataSourceExtractor;
+  public Statement build(final SingleStatementContext statement) {
+    final DataSourceExtractor dataSourceExtractor = new DataSourceExtractor(metaStore);
+    dataSourceExtractor.extractDataSources(statement);
 
-  AstBuilder(final DataSourceExtractor dataSourceExtractor) {
-    this.dataSourceExtractor = requireNonNull(dataSourceExtractor, "dataSourceExtractor");
+    final Node result = new Visitor(dataSourceExtractor, metaStore).visit(statement);
```

Comment:
How come passing both `dataSourceExtractor` and `metaStore` into the visitor is preferred to just passing `dataSourceExtractor` and calling `dataSourceExtractor.getMetaStore()`? (Trying to build up my understanding of Java best practices. Thanks!)

Reply:
Yes. Having an accessor to the data source extractor to get the metastore doesn't really make much sense on the `DataSourceExtractor` API.  The fact the class needs the metastore at all is sort of an internal implementation detail, so shouldn't be exposed in its API.  You should always try and keep the API of a type succinct, targeted and avoid exposing internal implementation details.

The visitor needs the metastore, so it should just be passed one.

Exposing one from the other increases coupling of the classes and makes them more complex to test. e.g. think of setting up the mocks for the Visitor. With the two separate parameters you just pass in two mocks, but with the old pattern you have to pass set up the first mock to return the second mock.